packages/nanocurrency/src/assembly/blake2b.ts

```diff
@@ -0,0 +1,181 @@
+];
+
+let ctx: Context = {
+  b: new Uint8Array(128),
+  h: new Uint64Array(8),
+  t: new Uint64Array(2), // input count
+  c: 0, // pointer within buffer
+  outlen: HASH_LENGTH, // output length in bytes
+};
+
+// Compression function. 'last' flag indicates last block.
+let v = new Uint64Array(16);
+let m = new Uint64Array(16);
+function blake2bCompress(last: boolean): void {
+  let i = 0;
+
+  v[0] = unchecked(ctx.h[0]);
+  v[8] = unchecked(BLAKE2B_IV[0]);
+  v[1] = unchecked(ctx.h[1]);
+  v[9] = unchecked(BLAKE2B_IV[1]);
+  v[2] = unchecked(ctx.h[2]);
+  v[10] = unchecked(BLAKE2B_IV[2]);
+  v[3] = unchecked(ctx.h[3]);
+  v[11] = unchecked(BLAKE2B_IV[3]);
+  v[4] = unchecked(ctx.h[4]);
+  v[12] = unchecked(BLAKE2B_IV[4]);
+  v[5] = unchecked(ctx.h[5]);
+  v[13] = unchecked(BLAKE2B_IV[5]);
+  v[6] = unchecked(ctx.h[6]);
+  v[14] = unchecked(BLAKE2B_IV[6]);
+  v[7] = unchecked(ctx.h[7]);
+  v[15] = unchecked(BLAKE2B_IV[7]);
+
+  v[12] = unchecked(v[12]) ^ unchecked(ctx.t[0]); // low 64 bits of offset
+  v[13] = unchecked(v[13]) ^ unchecked(ctx.t[1]); // high 64 bits
+
+  // last block flag set ?
+  if (last) {
+    v[14] = unchecked(~v[14]);
+  }
+
+  // get little-endian words
+  for (i = 0; i < 16; i++) {
+    m[i] = B2B_GET64(ctx.b, 8 * i);
+  }
+
+  // twelve rounds of mixing
+  for (i = 0; i < 12; i++) {
+    B2B_G(0, 4, 8, 12, unchecked(m[SIGMA[i][0]]), unchecked(m[SIGMA[i][1]]));
+    B2B_G(1, 5, 9, 13, unchecked(m[SIGMA[i][2]]), unchecked(m[SIGMA[i][3]]));
+    B2B_G(2, 6, 10, 14, unchecked(m[SIGMA[i][4]]), unchecked(m[SIGMA[i][5]]));
+    B2B_G(3, 7, 11, 15, unchecked(m[SIGMA[i][6]]), unchecked(m[SIGMA[i][7]]));
+    B2B_G(0, 5, 10, 15, unchecked(m[SIGMA[i][8]]), unchecked(m[SIGMA[i][9]]));
+    B2B_G(1, 6, 11, 12, unchecked(m[SIGMA[i][10]]), unchecked(m[SIGMA[i][11]]));
+    B2B_G(2, 7, 8, 13, unchecked(m[SIGMA[i][12]]), unchecked(m[SIGMA[i][13]]));
+    B2B_G(3, 4, 9, 14, unchecked(m[SIGMA[i][14]]), unchecked(m[SIGMA[i][15]]));
+  }
+
+  ctx.h[0] = unchecked(ctx.h[0]) ^ unchecked(v[0]) ^ unchecked(v[8]);
+  ctx.h[1] = unchecked(ctx.h[1]) ^ unchecked(v[1]) ^ unchecked(v[9]);
+  ctx.h[2] = unchecked(ctx.h[2]) ^ unchecked(v[2]) ^ unchecked(v[10]);
+  ctx.h[3] = unchecked(ctx.h[3]) ^ unchecked(v[3]) ^ unchecked(v[11]);
+  ctx.h[4] = unchecked(ctx.h[4]) ^ unchecked(v[4]) ^ unchecked(v[12]);
+  ctx.h[5] = unchecked(ctx.h[5]) ^ unchecked(v[5]) ^ unchecked(v[13]);
+  ctx.h[6] = unchecked(ctx.h[6]) ^ unchecked(v[6]) ^ unchecked(v[14]);
+  ctx.h[7] = unchecked(ctx.h[7]) ^ unchecked(v[7]) ^ unchecked(v[15]);
+}
+
+export function blake2bInit(): void {
+  // initialize hash state
+  ctx.h[0] = unchecked(BLAKE2B_IV[0]);
+  ctx.h[1] = unchecked(BLAKE2B_IV[1]);
+  ctx.h[2] = unchecked(BLAKE2B_IV[2]);
+  ctx.h[3] = unchecked(BLAKE2B_IV[3]);
+  ctx.h[4] = unchecked(BLAKE2B_IV[4]);
+  ctx.h[5] = unchecked(BLAKE2B_IV[5]);
+  ctx.h[6] = unchecked(BLAKE2B_IV[6]);
+  ctx.h[7] = unchecked(BLAKE2B_IV[7]);
+
+  ctx.h[0] ^= 0x01010000 ^ (0 << 8) ^ ctx.outlen;
```

Comment:
`(0 << 8)` is unnecessary. As I understand this was `keyLen << 8`? In this case you can just write `0x01010000 ^ ctx.outlen` or `0x01010000 ^ /*(0 << 8) ^*/ ctx.outlen`